File: plugin/zeppelin/src/main/scala/org/intellij/plugin/zeppelin/service/TableOutputHandler.scala
package org.intellij.plugin.zeppelin.service

import com.intellij.openapi.extensions.ExtensionPointName
import com.intellij.openapi.project.Project
import org.intellij.plugin.zeppelin.constants.ZeppelinConstants
import org.intellij.plugin.zeppelin.utils.ZeppelinLogger

trait TableOutputHandler {
  def invoke(project: Project, msg: String)
}

object TableOutputHandler {
  val Id = ZeppelinConstants.PLUGIN_ID + ".tableOutputHandler"
  val EP_NAME: ExtensionPointName[TableOutputHandler] =
    ExtensionPointName.create[TableOutputHandler](Id)

  def getAll: Array[TableOutputHandler] = EP_NAME.getExtensions
}


object DefaultTableOutputHandler extends TableOutputHandler {
  override def invoke(project: Project, msg: String): Unit = ZeppelinLogger.printMessage(msg)
}

File: zeppelin/implementation/src/org/intellij/plugin/zeppelin/extensionpoints/TableOutputHandler.kt
package org.intellij.plugin.zeppelin.extensionpoints

import com.intellij.openapi.extensions.ExtensionPointName
import com.intellij.openapi.project.Project
import org.intellij.plugin.zeppelin.constants.ZeppelinConstants
import org.intellij.plugin.zeppelin.utils.ZeppelinLogger

/**
 * Extension points to handle table outputs
 */
interface TableOutputHandler {
    fun handle(project: Project, msg: String)

    companion object {
        private const val ID: String = ZeppelinConstants.PLUGIN_ID + ".tableOutputHandler"
        private val EP_NAME: ExtensionPointName<TableOutputHandler> = ExtensionPointName.create<TableOutputHandler>(
                ID)

        fun getHandler(): TableOutputHandler = EP_NAME.extensions.firstOrNull()?:DefaultTableOutputHandler()
    }
}

class DefaultTableOutputHandler : TableOutputHandler {
    override fun handle(project: Project, msg: String): Unit = ZeppelinLogger.printMessage(msg)
}